packages/ckeditor5-engine/src/model/schema.ts

```diff
@@ -1014,6 +1015,65 @@ export default class Schema extends ObservableMixin() {
 			yield new Range( start, end );
 		}
 	}
+
+	/**
+	 * Returns a model range which is optimal (in terms of UX) for inserting a widget block.
+	 *
+	 * For instance, if a selection is in the middle of a paragraph, the collapsed range before this paragraph
+	 * will be returned so that it is not split. If the selection is at the end of a paragraph,
+	 * the collapsed range after this paragraph will be returned.
+	 *
+	 * Note: If the selection is placed in an empty block, the range in that block will be returned. If that range
+	 * is then passed to {@link module:engine/model/model~Model#insertContent}, the block will be fully replaced
+	 * by the inserted widget block.
+	 *
+	 * @param selection The selection based on which the insertion position should be calculated.
+	 * @param model Model instance.
+	 * @param place The place where to look for optimal insertion range.
+	 * The `auto` value will determine itself the best position for insertion.
+	 * The `before` value will try to find a position before selection.
+	 * The `after` value will try to find a position after selection.
+	 * @returns The optimal range.
+	 * @internal
+	 */
+	public findOptimalInsertionRange(
```

Comment:
I think we should not include `model` in the parameters list. Please have a look at other methods. This is in the engine package so we do not need to use public APIs, we can use internal APIs for `Range` and `Position` creation.